File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *detectCycle(struct ListNode *head) {
    struct ListNode *slow = head;
    struct ListNode *fast = head;
    bool has_cycle = false;
    
    while(!has_cycle && fast && fast->next){
        slow = slow->next;
        fast = fast->next->next;
        has_cycle = slow == fast;
    }
    if(!has_cycle){
        return false;
    }
    struct ListNode *p = head;
    while(p != slow){
        p = p->next;
        slow = slow->next;
    }
    return p;
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        slow = fast = head
        has_cycle = False
        while not has_cycle and fast and fast.next: 
            slow = slow.next
            fast = fast.next.next
            has_cycle = slow == fast
        if not has_cycle: 
            return None
        p = head
        while p != slow: 
            p = p.next
            slow = slow.next
        return p